File: backend/api/routes/screen_share.py
from flask import Blueprint, jsonify, request

from extensions import socketio
from services.screen_share_service import (
    create_screen_share_session,
    get_screen_share_session_by_room,
    end_screen_share_session,
)

bp = Blueprint("screen_share", __name__)


@bp.post("/api/screen-share/session")
def create_session():
    """
    Generate screen share room/session。
    """
    session = create_screen_share_session()

    base_url = request.host_url.rstrip("/")
    room_code = session["room_code"]

    return jsonify(
        {
            **session,
            "socket_url": base_url,
            "message": "screen share session created",
            "display_instruction": f"Display should join room {room_code}",
            "presenter_instruction": f"Presenter should join room {room_code}",
        }
    ), 201


@bp.get("/api/screen-share/session/<room_code>")
def get_session(room_code: str):
    """
    Check room session。
    """
    session = get_screen_share_session_by_room(room_code)

    if not session:
        return jsonify({"error": "screen share session not found"}), 404

    return jsonify(session)


@bp.post("/api/screen-share/session/<room_code>/end")
def end_session(room_code: str):
    """
    End screen share session。

    if session end for the first time：
    - update database
    - broadcast stop-screen-share

    if session ended before：
    - do not broadcast stop-screen-share again
    - return already_ended=True
    """
    session = end_screen_share_session(room_code)

    if not session:
        return jsonify({"error": "screen share session not found"}), 404

    already_ended = session.get("already_ended", False)

    if not already_ended:
        socketio.emit(
            "stop-screen-share",
            {
                "room": room_code,
                "reason": "session ended by backend",
            },
            to=room_code,
        )

        message = "screen share session ended"
    else:
        message = "screen share session was already ended"

    return jsonify(
        {
            **session,
            "message": message,
        }
    )

File: server/src/routes/screenShare.js
const express = require('express');
const { requireAuth, optionalAuth } = require('../middleware/auth');
const { asyncHandler } = require('../middleware/errors');
const {
  createSession,
  getSessionByRoom,
  endSession,
  getScreenState,
} = require('../services/screenShareService');

const router = express.Router();

// Controllers on LAN are not always admin-authenticated; allow create for presentation flow.
router.post(
  '/session',
  optionalAuth,
  asyncHandler(async (req, res) => {
    const session = createSession();
    const protocol = req.secure || req.app.get('secure') ? 'https' : req.protocol;
    const baseUrl = `${protocol}://${req.get('host')}`;
    res.status(201).json({
      ...session,
      socket_url: baseUrl,
      message: 'screen share session created',
      display_instruction: `Display should join room ${session.room_code}`,
      presenter_instruction: `Presenter should join room ${session.room_code}`,
    });
  })
);

router.get(
  '/session/:roomCode',
  asyncHandler(async (req, res) => {
    const session = getSessionByRoom(req.params.roomCode);
    if (!session) return res.status(404).json({ error: 'screen share session not found' });
    res.json(session);
  })
);

router.post(
  '/session/:roomCode/end',
  requireAuth,
  asyncHandler(async (req, res) => {
    const session = endSession(req.params.roomCode);
    if (!session) return res.status(404).json({ error: 'screen share session not found' });

    const io = req.app.get('io');
    if (io && !session.already_ended) {
      io.to(req.params.roomCode).emit('stop-screen-share', {
        room: req.params.roomCode,
        reason: 'session ended by backend',
      });
    }

    res.json({
      ...session,
      message: session.already_ended
        ? 'screen share session was already ended'
        : 'screen share session ended',
    });
  })
);

module.exports = router;
